### PR TITLE
1030 hi l1a update sci de packet definition and parsing

### LINKED ISSUE #1030: Hi L1A - Update SCI_DE packet definition and parsing
### Algorithm Description:

From Hi Tag-up meeting, Paul provided the following:

- Meta-events
   - Should be considered as part of header of SCI_DE packet (2 bit field does not designate meta-event)
   - 2 SCI_DE packets per ESA step - both meta-events are expected to have the same time-tag
- Time-tag for a DE is at the start (first 16-bits)
- Time tags for when a DE happened are ticking at 1999 us according to the engineering LUT.  Not sure why it's not 2000 us.  It's a parameter to care about.
- This is an instrument status summary item. It is up to Tim's discretion as to how hard it is to implement code to have a dynamic value.

### PR DESCRIPTION
**Pull request by tmplummer**
# Change Summary

## Overview
These changes are driven by an update to the Hi SCI_DE packet definition. Meta-events have been pulled out of the DE_TOF binary blob and are present in every packet. Additionally, `last_spin_num` and `spin_invalids` bit-flag have been added.

## Updated Files
- imap_processing/cdf/config/imap_hi_variable_attrs.yaml
   - Added the following variables to hi_l1a_de:
      - `src_seq_ctr`
      - `pkt_len`
      - `last_spin_num`
      - `spin_invalids`
- imap_processing/hi/l1a/hi_l1a.py
   - Point to updated xtce packet definition file
- imap_processing/hi/l1a/science_direct_event.py
   - Refactored much of this code due to change in packet definition.
   - Meta-event is no longer in DE_TOF binary blob
   - Vectorized calculation of DE MET
- imap_processing/hi/packet_definitions/TLM_HI_COMBINED_SCI.xml
   - Updates to packet definition: Meta-event fields are no longer in the DE_TOF binary blob
- imap_processing/tests/hi/test_science_direct_event.py
   - Despite the diff, the file imap_processing/tests/hi/test_l1a_sci_de.py was renamed and the test updated.

Closes: #1030

### CODE REVIEW

File: imap_processing/cdf/config/imap_hi_variable_attrs.yaml [hunk 1]
```diff
@@ -126,12 +126,33 @@ hi_esa_step: &hi_esa_step
...
+
+hi_de_pkt_len:
+  <<: *default_ccsds_pkt_len
+
 hi_de_ccsds_met:
   <<: *hi_ccsds_met
 
+hi_de_last_spin_num:
+  <<: *default_uint8
+  CATDESC: Spin number (1-based) for the first spin in the packet
```
Comment (tech3371) [inline imap_processing/cdf/config/imap_hi_variable_attrs.yaml:140]: what does `(1-based)` mean?

File: imap_processing/hi/l1a/science_direct_event.py [hunk 2]
```diff
@@ -12,69 +15,39 @@
...
-    |        .....
-    |    ]
-    |    second packet = [
-    |        (start_bitmask_data, tof_1, tof_2, tof_3, de_tag),
-    |       .....
-    |    ]
+    There should be two data packets per ESA. Each packet contains meta-event
+    data that is identical between the two packets for a common ESA.
+    If there is no event record for certain ESA step, then both packets will
+    contain 0-bits in DE_TOF.
```
Comment (tech3371) [inline imap_processing/hi/l1a/science_direct_event.py:50]: good to know!

File: imap_processing/hi/l1a/science_direct_event.py [hunk 3]
```diff
@@ -93,45 +66,26 @@ def parse_direct_event(event_data: str) -> dict:
```
Comment (tech3371) [inline imap_processing/hi/l1a/science_direct_event.py:66]: We will need to update line 64's comment

File: imap_processing/hi/l1a/science_direct_event.py [hunk 4]
```diff
@@ -321,23 +202,32 @@ def science_direct_event(packets_data: xr.Dataset) -> xr.Dataset:
...
+            "src_seq_ctr": "src_seq_ctr",
+            "pkt_len": "pkt_len",
+            "last_spin_num": "last_spin_num",
+            "spin_invalids": "spin_invalids",
+            "esa_step_num": "esa_step",
+            "meta_seconds": "meta_seconds",
+            "meta_subseconds": "meta_subseconds",
+        }.items():
+            de_data_dict[to_key].extend(
+                [packets_data[from_key].data[i]] * len(parsed_de_data["de_tag"])
```
Comment (tech3371) [inline imap_processing/hi/l1a/science_direct_event.py:231]: What does this line do? Just curious

File: imap_processing/hi/packet_definitions/TLM_HI_COMBINED_SCI.xml [hunk 5]
```diff
@@ -1,6 +1,6 @@
 <?xml version='1.0' encoding='utf-8'?>
 <xtce:SpaceSystem xmlns:xtce="http://www.omg.org/space/xtce" name="H45">
-	<xtce:Header date="2024-11-21 00:00:00" version="1" author="IMAP SDC" />
+	<xtce:Header date="2024-11-26 00:00:00" version="1" author="IMAP SDC" source_file="TLM_H45_H90_20241126.xlsx" />
```
Comment (tech3371) [inline imap_processing/hi/packet_definitions/TLM_HI_COMBINED_SCI.xml:3]: nice! I think we all should store source file

File: imap_processing/tests/hi/test_science_direct_event.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
+from imap_processing.hi.l1a.science_direct_event import create_dataset
+
+
+def test_create_dataset():
+    """Test create_dataset"""
+    # dummy data to test create_dataset
+    # dummy data to test create_dataset
```
Comment (tech3371) [inline imap_processing/tests/hi/test_science_direct_event.py]: remove duplicate?

**Review by tech3371 — APPROVED** [review]
Minor questions. All non-blockers. Makes sense to me and LGTM

File: imap_processing/cdf/config/imap_hi_variable_attrs.yaml (hunk 1 above)
Comment (tmplummer) [reply imap_processing/cdf/config/imap_hi_variable_attrs.yaml:140]: It just means that there is no zeroth spin number, it starts at 1.

File: imap_processing/hi/l1a/science_direct_event.py (hunk 4 above)
Comment (tmplummer) [reply imap_processing/hi/l1a/science_direct_event.py:231]: Haha, yeah. I can comment that. It is actually a remnant from what you wrote originally. It just makes a list of identical values that is the length of the number of parsed DEs.

File: imap_processing/hi/packet_definitions/TLM_HI_COMBINED_SCI.xml (hunk 5 above)
Comment (tmplummer) [reply imap_processing/hi/packet_definitions/TLM_HI_COMBINED_SCI.xml:3]: I can put that in a PR change to `excel_to_xtce`.

**Pull request merged**